Fix multiple installersets for tektonaddon

savitaashture opened the pull request:
# Changes

This PR fixes creation of multiple installer set for clustertasks addon


<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

<!--
Describe any user facing changes here, or delete this block.

Examples of user facing changes:
- API changes
- Bug fixes
- Any changes in behavior
- Changes requiring upgrade notices or deprecation warnings

For pull requests with a release note:

```release-note
Your release note here
```

For pull requests that require additional action from users switching to the new release, include the string "action required" (case insensitive) in the release note:

```release-note
action required: your release note here
```

For pull requests that don't need to be mentioned at release time, use the `/release-note-none` Prow command to add the `release-note-none` label to the PR. You can also write the string "NONE" as a release note in your PR description:

```release-note
NONE
```
-->
```release-note
NONE
```

pradeepitm12 commented on pkg/reconciler/openshift/tektonaddon/tektonaddon.go:
```diff
@@ -416,41 +420,27 @@ func createInstallerSet(ctx context.Context, oc clientset.Interface, ta *v1alpha
-				tektoninstallerset.CreatedByKey: createdByValue,
+				tektoninstallerset.CreatedByKey:      createdByValue,
+				tektoninstallerset.InstallerSetType:  component,
+				tektoninstallerset.ReleaseVersionKey: releaseVersion,
```
:+1:

vdemeester reviewed (commented):
@savitaashture can we have some tests ? (not sure how easy it would be but still)

savitaashture commented:
> @savitaashture can we have some tests ? (not sure how easy it would be but still)

Updated e2e test to verify installerSets for addon based on label

sm43 commented on pkg/reconciler/kubernetes/tektoninstallerset/const.go:
```diff
@@ -21,4 +21,12 @@ const (
+	TriggersResourcesInstallerSet      = "TriggersResources"
+	ConsoleCLIInstallerSet             = "ConsoleCLI"
+	MiscellaneousResourcesInstallerSet = "MiscellaneousResources"
+	CreatedByValue                     = "TektonAddon"
```
these are addon specific const right?
should we keep them in addon pkg? and not in tektoninstallerset?

sm43 commented on pkg/reconciler/openshift/tektonaddon/tektonaddon.go:
```diff
@@ -369,26 +354,34 @@ func (r *Reconciler) ensureClusterTasks(ctx context.Context, ta *v1alpha1.Tekton
-		ctIs, err := oc.OperatorV1alpha1().TektonInstallerSets().
-			Get(ctx, compInstallerSet, metav1.GetOptions{})
+	// Delete installer sets if exist more than one
+	if len(installerSets.Items) > 1 {
```
here if there are more than 1 installer set we are deleting all right?
then do we need to continue after deleting?

as after deleting
we are looping on items
```
for i := range installerSets.Items {
```
but we have already deleted all

sm43 replied:
we need a return statement I think after deleting

sm43 commented on pkg/reconciler/openshift/tektonaddon/tektonaddon.go:
```diff
@@ -369,26 +354,34 @@ func (r *Reconciler) ensureClusterTasks(ctx context.Context, ta *v1alpha1.Tekton
 // and if installer set which already exist is of older version then it deletes and return false to create a new
 // installer set
 func checkIfInstallerSetExist(ctx context.Context, oc clientset.Interface, relVersion string,
```
as this is more generalised now, can we move this to installer Set pkg
so that other component can use the same?
something like `installerSet.GetByLabel()`

nikhil-thomas replied:
separate pr sound fine. 
could you create and issue (or jira) to track this.

nikhil-thomas commented on pkg/reconciler/openshift/tektonaddon/tektonaddon.go:
```diff
@@ -369,26 +354,34 @@ func (r *Reconciler) ensureClusterTasks(ctx context.Context, ta *v1alpha1.Tekton
 // installer set
 func checkIfInstallerSetExist(ctx context.Context, oc clientset.Interface, relVersion string,
-	ta *v1alpha1.TektonAddon, component string) (bool, error) {
+	component string) (bool, error) {
```
the instead of sending the component, version etc separately the last paramerte should be `labelSelector` so that whoever using this function could format `1` or `n` labels and pass it to this function. 

Otherwise, we will have to change the the implementation of this function for each use case for other components in the future.

nikhil-thomas commented on pkg/reconciler/openshift/tektonaddon/tektonaddon.go:
```diff
@@ -416,41 +410,27 @@ func createInstallerSet(ctx context.Context, oc clientset.Interface, ta *v1alpha
 			GenerateName: fmt.Sprintf("%s-", prefix),
 			Labels: map[string]string{
-				tektoninstallerset.CreatedByKey: createdByValue,
+				tektoninstallerset.CreatedByKey:      tektoninstallerset.CreatedByValue,
```
`tektoninstallerset.CreatedByValue` this definitely a reconciler specific item, which has to live in the reconciler which is using this (for example pipelines, addons, triggers..)(

nikhil-thomas replied:
what i meant is `tektoninstallerset.CreatedByValue` sounds odd, i was expecting just `CreatedByValue` (as CreatedByValue is defined in addons. 

In other words, the value part of this annotation comes from the entity who creates the installerset. So referencing it as `tektoninstallerset.` sounds odd.

nikhil-thomas commented on pkg/reconciler/openshift/tektonaddon/tektonaddon.go:
```diff
@@ -416,41 +410,27 @@ func createInstallerSet(ctx context.Context, oc clientset.Interface, ta *v1alpha
 }
 
-func makeInstallerSet(ta *v1alpha1.TektonAddon, manifest mf.Manifest, prefix, releaseVersion, specHash string) *v1alpha1.TektonInstallerSet {
+func makeInstallerSet(ta *v1alpha1.TektonAddon, manifest mf.Manifest, prefix, releaseVersion, component, specHash string) *v1alpha1.TektonInstallerSet {
```
here also let us pass the set of labes as a single argument (label selector itself or map). 
so that formatiing the labels will be the responsibility of the caller.

nikhil-thomas commented on pkg/reconciler/openshift/tektonaddon/tektonaddon.go:
```diff
@@ -462,28 +442,14 @@ func makeInstallerSet(ta *v1alpha1.TektonAddon, manifest mf.Manifest, prefix, re
 
-func (r *Reconciler) deleteInstallerSet(ctx context.Context, ta *v1alpha1.TektonAddon, component string) error {
+func (r *Reconciler) deleteInstallerSet(ctx context.Context, component string) error {
 
```
same here, instead of the `component string` let us pass label selector as the last param

nikhil-thomas commented on pkg/reconciler/openshift/tektonaddon/tektonaddon.go:
```diff
@@ -462,28 +442,14 @@ func makeInstallerSet(ta *v1alpha1.TektonAddon, manifest mf.Manifest, prefix, re
-			return err
-		}
+	err := r.operatorClientSet.OperatorV1alpha1().TektonInstallerSets().
+		DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{
```
👍

nikhil-thomas commented on pkg/reconciler/openshift/tektonaddon/tektonaddon.go:
```diff
@@ -369,26 +354,34 @@ func (r *Reconciler) ensureClusterTasks(ctx context.Context, ta *v1alpha1.Tekton
-		ctIs, err := oc.OperatorV1alpha1().TektonInstallerSets().
-			Get(ctx, compInstallerSet, metav1.GetOptions{})
+	// Delete installer sets if exist more than one
+	if len(installerSets.Items) > 1 {
```
why is this `>` 1 we should delete all, shouldn't we.

nikhil-thomas replied:
in other words, we as @sm43 said we should return if we are deleting all. 

I think this `> 1 ` check is misleading. We should stick to a single mechanism irrespective of the number of instalerSets returned (delete 1 or many).

nikhil-thomas commented on pkg/reconciler/openshift/tektonaddon/tektonaddon.go:
```diff
@@ -369,26 +354,34 @@ func (r *Reconciler) ensureClusterTasks(ctx context.Context, ta *v1alpha1.Tekton
 
-		version, ok := ctIs.Annotations[tektoninstallerset.ReleaseVersionKey]
+	// Check if installer set is already created
+	for i := range installerSets.Items {
```
what is the purpose of this `for loop`. 
the above `DeleteCollection()` call will send a `delete `  api call so why do we try to process the same set with this `for loop`.

All the items in this  instalerset.Items is being deleted because of the above call, isn't it.

nikhil-thomas replied:
the DeleteCollection call only makes an Api call, it won't make `installerSets.Items` empty.

savitaashture commented:
@sm43 @nikhil-thomas I have addressed review comments except refactoring part 
Planning to do refactor  for `checkIfInstallerSetExist` `createInstallerSet` and  `makeInstallerSet` common for all 
as a separate PR once @pradeepitm12 changes are also in for this multiple installerset fix for Pipeline and Triggers

WDYT?

sm43 commented on pkg/reconciler/openshift/tektonaddon/extension.go:
```diff
@@ -134,7 +131,7 @@ func (oe openshiftExtension) PostReconcile(ctx context.Context, comp v1alpha1.Te
 
 	// spec hash stored on installerSet
-	lastAppliedHash := installedTIS.GetAnnotations()[tektoninstallerset.LastAppliedHashKey]
+	lastAppliedHash := installedTIS.Items[0].GetAnnotations()[tektoninstallerset.LastAppliedHashKey]
```
should we have a check for `len` for Items ? just to be safe before accessing

savitaashture commented on pkg/reconciler/openshift/tektonaddon/extension.go:
```diff
@@ -120,21 +117,25 @@ func (oe openshiftExtension) PostReconcile(ctx context.Context, comp v1alpha1.Te
 		return err
 	}
 
+	if len(installedTIS.Items) == 0 {
+		return v1alpha1.RECONCILE_AGAIN_ERR
+	}
+
```
@sm43 if len is 0 shall we return Reconcile again and wait for sometime or shall we return error itself

nikhil-thomas replied:
https://github.com/tektoncd/operator/pull/574/files#r782647681

nikhil-thomas replied:
adding a explicit wait in situations like these is a kubernetes anti-pattern.

if the cluster state is not " yet " the desired state, we exit out of the reconcile loop (return RECONCILE_AGAIN) and let the future loops proceed past the check.

nikhil-thomas commented on pkg/reconciler/openshift/tektonaddon/tektonaddon.go:
```diff
@@ -369,40 +358,36 @@ func (r *Reconciler) ensureClusterTasks(ctx context.Context, ta *v1alpha1.Tekton
-		if err != nil {
-			return false, err
-		}
+	// release version doesn't exist or is different from expected
+	// deleted existing InstallerSet and create a new one
```
```suggestion
	// release version doesn't exist or is different from expected
	// deleted existing InstallerSet and create a new one
	// or there is more than one installerset (unexpected)
```

nikhil-thomas commented on pkg/reconciler/openshift/tektonaddon/tektonaddon.go:
```diff
@@ -369,40 +358,36 @@ func (r *Reconciler) ensureClusterTasks(ctx context.Context, ta *v1alpha1.Tekton
+	// release version doesn't exist or is different from expected
+	// deleted existing InstallerSet and create a new one
+	if err = oc.OperatorV1alpha1().TektonInstallerSets().
+		DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{
+			LabelSelector: labelSelector,
+		}); err != nil {
+		return false, err
```
this should work but ideally, we should loop through the list and call individual delete() after ensuring `DeletionTimestamp` is not set in metadata of each installerset. 

 We could be here even after a previous reconcile loop has called `DeleteCollection` as the installerset might take few seconds to clean up (finalize) what it has created.

nikhil-thomas replied:
ie, the DeletionCollection() is just an API call, so we cannot gurantee when the installerSet will be completely removed.

Mentioning this here, to make sure that we are all on the same page.
cc @sm43

nikhil-thomas commented on pkg/reconciler/openshift/tektonaddon/extension.go:
```diff
@@ -120,21 +117,25 @@ func (oe openshiftExtension) PostReconcile(ctx context.Context, comp v1alpha1.Te
 
 			if err := createInstallerSet(ctx, oe.operatorClientSet, addon, manifest, oe.version,
-				miscellaneousResourcesInstallerSet, "addon-openshift"); err != nil {
+				MiscellaneousResourcesInstallerSet, "addon-openshift"); err != nil {
```
retrun `RCONCILE_AGAIN_ERR` here to ensure that this API call has created the installerSet.

nikhil-thomas commented on pkg/reconciler/openshift/tektonaddon/extension.go:
```diff
@@ -120,21 +117,25 @@ func (oe openshiftExtension) PostReconcile(ctx context.Context, comp v1alpha1.Te
 		return err
 	}
 
+	if len(installedTIS.Items) == 0 {
+		return v1alpha1.RECONCILE_AGAIN_ERR
+	}
```
as creating a new installerset will not have any effect on the length of current installedTIS.items, let us move this return to https://github.com/tektoncd/operator/pull/574/files#r782647382

nikhil-thomas commented:
@savitaashture could you make 2 changes
1. https://github.com/tektoncd/operator/pull/574/files#r782647681
2. (minor comment which will help future readers 🧑‍💻 ) https://github.com/tektoncd/operator/pull/574#discussion_r782641306

savitaashture commented:
> @savitaashture could you make 2 changes
> 
> 1. https://github.com/tektoncd/operator/pull/574/files#r782647681
> 2. (minor comment which will help future readers technologist ) [Fix multiple installersets for tektonaddon #574 (comment)](https://github.com/tektoncd/operator/pull/574#discussion_r782641306)

@nikhil-thomas Addressed :+1:

nikhil-thomas commented on pkg/reconciler/openshift/tektonaddon/extension.go:
```diff
@@ -87,31 +87,28 @@ func (oe openshiftExtension) PostReconcile(ctx context.Context, comp v1alpha1.Te
-			miscellaneousResourcesInstallerSet, "addon-openshift"); err != nil {
-			return err
+			MiscellaneousResourcesInstallerSet, "addon-openshift"); err != nil {
+			return v1alpha1.RECONCILE_AGAIN_ERR
```
if err is nil then return `v1alpha1.RECONCILE_AGAIN_ERR` 
else return err.. 

otherwise we will never catch the error

savitaashture replied:
:woman_facepalming:  :see_no_evil:

nikhil-thomas commented:
✅ lgtm

nikhil-thomas commented:
/approve

sm43 commented:
LGTM
